src/vm/internal/otable.cc

```diff
@@ -1,271 +1,148 @@
+ObjectTable::Value ObjectTable::find(Key const & key) {
+    auto it = objects_.find(key);
+    if( it != objects_.end() ) {
+        return it->second;
+    } 
+    else {
+        return nullptr;
     }
-    curr = curr->next_ch_hash;
-  }
-  FREE_MSTR(s);
-  vec = resize_array(vec, count);
-  return (vec);
 }
 
-/*
- * Add an object to the table - can't have duplicate names.
- *
- * Exception: Precompiled objects have a dummy entry here, but it is
- * guaranteed to be behind the real entry if a real entry exists.
- */
-
-static int objs_in_table = 0;
-
-void enter_object_hash(object_t *ob) {
-#ifdef DEBUG
-  object_t *s;
-#endif
-  char *base;
-  int dummy;
-#ifndef DEBUG
-  h = ObjHash(ob->obname);
-#else
-  s = find_obj_n(ob->obname); /* This sets h */
-  /* when these reload, the new copy comes in before the old goes out */
-  if (s != master_ob && s != simul_efun_ob) {
-    DEBUG_CHECK1(s && s != ob, "Duplicate object \"/%s\" in object hash table", ob->obname);
-  }
-#endif
-
-  ob->next_hash = obj_table[h];
-  obj_table[h] = ob;
-  objs_in_table++;
-
-  // for children()
-  base = _basename(ob->obname, &dummy);
-
-  ch = ObjHash(base);
-  ob->next_ch_hash = ch_table[ch];
-  ch_table[ch] = ob;
-  FREE_MSTR(base);
-  return;
+//return the children of key if any as a vector of object pointers(Values).
+ObjectTable::Vector ObjectTable::children(Key const & key) {
+    auto it = children_.find( basename(key) );
+    if( it != children_.end() ) {
+        return it->second;
+    }
+    else {
+        return Vector({});
+    }
 }
 
-/*
- * Remove an object from the table - generally called when it
- * is removed from the next_all list - i.e. in destruct.
- */
-
-void remove_object_hash(object_t *ob) {
-  int dummy;
-  object_t *s;
-  object_t *t = 0;
-  char *base;
-
-  s = find_obj_n(ob->obname); /* this sets h, and cycles the ob to the front */
-  if (!s) {
-    fatal("couldn't find object %s in obj_table", ob->obname);
-  }
-
-  DEBUG_CHECK1(s != ob, "Remove object \"/%s\": found a different object!", ob->obname);
-
-  obj_table[h] = ob->next_hash;
-  ob->next_hash = 0;
-  objs_in_table--;
-
-  base = _basename(ob->obname, &dummy);
-  ch = ObjHash(base);
-  FREE_MSTR(base);
-  s = ch_table[ch];
-  while (s && s != ob) {
-    t = s;
-    s = s->next_ch_hash;
-  }
-  if (!s) {
-    fatal("object not found in children list");
-  }
-  if (!t) {
-    ch_table[ch] = s->next_ch_hash;
-  } else {
-    t->next_ch_hash = s->next_ch_hash;
-  }
-  ob->next_ch_hash = 0;
-  return;
+//Attempt to find the object with name key in the object table and remove it
+//from both the objects map and the children map if found. Returns true if found,
+//otherwise false.
+//TODO: generalize this code by removing dependency on field obname.
+bool ObjectTable::remove(Key const & key) {
+    
+    auto it1 = objects_.find(key);
+    if( it1 != objects_.end() ) {
```

Comment:
you can reverse the if case and make it more clear

``` if(it1==objcets_.end()) { return false; }``` 

Also, it2 and it3 is what? they are just a variety of ijk without real meanings.